src/idiomatic/leveraging-the-type-system/raii.md

```diff
@@ -0,0 +1,116 @@
+---
+minutes: 30
+---
+
+# RAII and `Drop` in Practice
+
+RAII (_Resource Acquisition Is Initialization_) means tying the lifetime of a
+resource to the lifetime of a value.
+
+Rust applies RAII automatically for memory management. The `Drop` trait lets you
+extend this pattern to anything else.
+
+```rust
+use std::sync::Mutex;
+
+fn main() {
+    let mux = Mutex::new(vec![1, 2, 3]);
+
+    {
+        let mut data = mux.lock().unwrap();
+        data.push(4); // lock held here
+    } // lock automatically released here
+}
+```
+
+<details>
+
+- In the above example
+  [the `Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html) owns its
+  data: you can’t access the value inside without first acquiring the lock.
+
+  `mux.lock()` returns a
+  [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html),
+  which [dereferences](https://doc.rust-lang.org/std/ops/trait.DerefMut.html) to
+  the data and implements
+  [`Drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html).
+
+- You may recall from
+  [the Memory Management chapter](../../memory-management/drop.md) that the
+  [`Drop` trait](https://doc.rust-lang.org/std/ops/trait.Drop.html) lets you
+  define what should happen when a resource is dropped.
+
+  - In
+    [the Blocks and Scopes chapter](../../control-flow-basics/blocks-and-scopes.md),
+    we saw the most common situation where a resource is dropped: when the scope
+    of its _owner_ ends at the boundary of a block (`{}`).
+
+  - The use of
+    [`std::mem::drop(val)`](https://doc.rust-lang.org/std/mem/fn.drop.html)
+    allows you to _move_ a value out of scope before the block ends.
+
+  - There are also other scenarios where this can happen, such as when the value
+    owning the resource is "shadowed" by another value:
+
+    ```rust
+    let a = String::from("foo");
+    let a = 3; // ^ The previous string is dropped here
+               //   because we shadow its binding with a new value.
+    ```
+
+  - Recall also from [the Drop chapter](../../memory-management/drop.md) that
+    for a composite type such as a `struct`, all its fields will be dropped when
+    the struct itself is dropped. If a field implements the `Drop` trait, its
+    `Drop::drop` _trait_ method will also be invoked.
+
+- In any scenario where the stack unwinds the value, it is guaranteed that the
```

Comment:
The stack doesn't unwind things -- the stack _is_ unwound in handling a panic or normal function return.